apps/managedidentity/managedidentity.go

```diff
@@ -76,62 +93,155 @@ func (s systemAssignedValue) value() string { return string(s) }
+func WithHTTPClient(httpClient ops.HTTPClient) ClientOption {
+	return func(o *ClientOptions) {
+		o.httpClient = httpClient
+	}
 }
 
 // Client to be used to acquire tokens for managed identity.
 // ID: [SystemAssigned()], [ClientID("clientID")], [ResourceID("resourceID")], [ObjectID("objectID")]
 //
 // Options: [WithHTTPClient]
-func New(id ID, options ...Option) (Client, error) {
-	fmt.Println("idType: ", id.value())
-
-	opts := clientOptions{
-		claims: "claims",
+func New(id ID, options ...ClientOption) (Client, error) {
+	opts := ClientOptions{
+		httpClient: shared.DefaultClient,
 	}
 
 	for _, option := range options {
-		option.apply(&opts)
+		option(&opts)
 	}
 
-	authInfo, err := authority.NewInfoFromAuthorityURI("authorityURI", true, false)
+	client := Client{
+		miType:     id,
+		httpClient: opts.httpClient,
+	}
+
+	return client, nil
+}
+
+func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
+	var msiEndpoint *url.URL
+	msiEndpoint, err := url.Parse(imdsEndpoint)
 	if err != nil {
-		return Client{}, err
+		return nil, errors.New("error creating URL as parsing the URL filed")
 	}
 
-	authParams := authority.NewAuthParams(id.value(), authInfo)
-	client := Client{ // Note: Hey, don't even THINK about making Base into *Base. See "design notes" in public.go and confidential.go
-		AuthParams:      authParams,
-		cacheAccessorMu: &sync.RWMutex{},
-		// manager:         storage.New(token),
-		// pmanager:        storage.NewPartitionedManager(token),
+	msiParameters := msiEndpoint.Query()
+	msiParameters.Add(apiVersionQuerryParameterName, "2018-02-01")
+
+	resource = strings.TrimSuffix(resource, "/.default")
+
+	msiParameters.Add(resourceQuerryParameterName, resource)
+
+	if len(claims) > 0 {
+		msiParameters.Add("claims", claims)
+	}
+
+	switch t := id.(type) {
+	case ClientID:
+		if len(string(t)) > 0 {
+			msiParameters.Add(miQuerryParameterClientId, string(t))
+		} else {
+			return nil, fmt.Errorf("clientId parameter is empty for %T", t)
+		}
+	case ResourceID:
+		if len(string(t)) > 0 {
+			msiParameters.Add(miQuerryParameterResourceId, string(t))
+		} else {
+			return nil, fmt.Errorf("resourceID parameter is empty for %T", t)
+		}
+	case ObjectID:
+		if len(string(t)) > 0 {
+			msiParameters.Add(miQuerryParameterObjectId, string(t))
+		} else {
+			return nil, fmt.Errorf("objectID parameter is empty for %T", t)
+		}
+	case systemAssignedValue: // not adding anything
+	default:
+		return nil, fmt.Errorf("unsupported type %T", id)
 	}
 
-	return client, err
+	msiEndpoint.RawQuery = msiParameters.Encode()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, msiEndpoint.String(), nil)
+	if err != nil {
+		return nil, errors.New("error creating request")
+	}
+	return req, nil
+}
+
+func getTokenForRequest(req *http.Request, httpClient ops.HTTPClient) (accesstokens.TokenResponse, error) {
+	req.Header.Add(metaHTTPHeadderName, "true")
```

Comment:
Logically this one belongs to the request creator function above.

Reply:
👍 not all APIs expect this header

Reply:
Updated this added this for `createIMDSAuthRequest`